### PR TITLE
Add no-inner-declarations rule

### PR DESCRIPTION
**Pull request by bartlomieju**
Before landing we must verify that it yields the same results in `std/` as ESLint.

CC @magurotuna would you be interested in integrating?

### CODE REVIEW

**Comment by magurotuna** [comment]
@bartlomieju Yes, I'll check it.

File: src/rules/no_inner_declarations.rs [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+      "no-inner-declarations",
+      &format!("Move {} declaration to {} root", kind, root),
+    );
+  }
+}
+
+impl Visit for NoInnerDeclarationsVisitor {
+  noop_visit_type!();
+
+  fn visit_function(&mut self, fn_: &ast::Function, parent: &dyn Node) {
```
Comment (magurotuna) [inline src/rules/no_inner_declarations.rs:171]: Seems that implementing `visit_arrow_expr` here is also necessary to check if it's inside function

Comment (bartlomieju) [reply src/rules/no_inner_declarations.rs:171]: Ah yeah, good catch 👍

Comment (bartlomieju) [reply src/rules/no_inner_declarations.rs:178]: Done

**Comment by magurotuna** [comment]
Great!
As I said at https://github.com/denoland/deno/issues/7285#issuecomment-683843792, now we get the same result as ESLint.

**Review by magurotuna — APPROVED** [review]
LGTM :+1:

**Pull request merged**